src/main/cpp/win_fsnotifier.cpp

```diff
@@ -0,0 +1,209 @@
+// Initial version copied from Initial version copied from https://github.com/JetBrains/intellij-community/blob/master/native/WinFsNotifier/fileWatcher3.c
+// Copyright 2000-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
+
+#ifdef _WIN32
+
+#include "native.h"
+#include "generic.h"
+#include "win.h"
+
+JavaVM* jvm = NULL;
+
+typedef struct watch_details {
+    HANDLE threadHandle;
+    HANDLE stopEventHandle;
+    wchar_t drivePath[4];
+    int watchedPathCount;
+    wchar_t **watchedPaths;
+    jobject watcherCallback;
+} watch_details_t;
+
+// TODO Find the right size for this
+#define EVENT_BUFFER_SIZE (16*1024)
+
+#define CREATE_SHARE (FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE)
+#define CREATE_FLAGS (FILE_FLAG_BACKUP_SEMANTICS | FILE_FLAG_OVERLAPPED)
+#define EVENT_MASK (FILE_NOTIFY_CHANGE_FILE_NAME | FILE_NOTIFY_CHANGE_DIR_NAME | \
+                    FILE_NOTIFY_CHANGE_ATTRIBUTES | FILE_NOTIFY_CHANGE_SIZE | FILE_NOTIFY_CHANGE_LAST_WRITE)
+
+void handlePathChanged(watch_details_t *details, FILE_NOTIFY_INFORMATION *info) {
+    // const char *event;
+    // if (info->Action == FILE_ACTION_ADDED || info->Action == FILE_ACTION_RENAMED_OLD_NAME) {
+    //     event = "CREATE";
+    // } else if (info->Action == FILE_ACTION_REMOVED || info->Action == FILE_ACTION_RENAMED_OLD_NAME) {
+    //     event = "DELETE";
+    // } else if (info->Action == FILE_ACTION_MODIFIED) {
+    //     event = "CHANGE";
+    // } else {
+    //     return;  // unknown event
+    // }
+
+    int pathLen = info->FileNameLength / sizeof(wchar_t);
+    wchar_t *changedPath = add_prefix(info->FileName, pathLen, details->drivePath);
+
+    bool watching = false;
+    for (int i = 0; i < details->watchedPathCount; i++) {
+        wchar_t* watchedPath = details->watchedPaths[i];
+        if (wcsncmp(watchedPath, changedPath, wcslen(watchedPath)) == 0) {
```

Comment:
I wonder if we need to do something regarding case sensitivity here. Or the caller needs to make sure the watched path has the right case.

Reply:
Good point.

Reply:
We need to do the same on macOS as well I guess.

Reply:
...at least as testing goes.

Reply:
Added test for case differences.